src/windows/mod.rs

```diff
@@ -0,0 +1,390 @@
+            )
+        });
+        SegmentIter {
+            sections: sections.unwrap_or(&[][..]),
+            phantom: PhantomData,
+        }
+    }
+
+    #[inline]
+    fn virtual_memory_bias(&self) -> Bias {
+        Bias(self.nt_headers().map_or(0, |nt_headers| {
+            nt_headers.OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_DEBUG as usize]
+                .VirtualAddress as isize
+        }))
+    }
+
+    fn each<F, C>(mut f: F)
+    where
+        F: FnMut(&Self) -> C,
+        C: Into<IterationControl>,
+    {
+        let proc = unsafe { GetCurrentProcess() };
+        let mut modules_size = 0;
+        unsafe {
+            if EnumProcessModules(proc, ptr::null_mut(), 0, &mut modules_size) == 0 {
+                return;
+            }
+        }
+        let module_count = modules_size / mem::size_of::<HMODULE>() as u32;
+        let mut modules = vec![unsafe { mem::zeroed() }; module_count as usize];
+        unsafe {
+            if EnumProcessModules(proc, modules.as_mut_ptr(), modules_size, &mut modules_size) == 0
+            {
+                return;
+            }
+        }
+
+        modules.truncate(modules_size as usize / mem::size_of::<HMODULE>());
+
+        for module in modules.iter_mut() {
+            unsafe {
+                let mut module_path = vec![0u16; MAX_PATH + 1];
+                if GetModuleFileNameExW(
+                    proc,
+                    *module,
+                    module_path.as_mut_ptr(),
+                    MAX_PATH as u32 + 1,
+                ) == 0
+                {
+                    continue;
+                }
+
+                let mut module_info = mem::zeroed();
+                if !GetModuleInformation(
+                    proc,
+                    *module,
+                    &mut module_info,
+                    mem::size_of::<MODULEINFO>() as u32,
+                ) == 0
+                {
+                    continue;
+                }
+
+                // to prevent something else from unloading the module while
+                // we're poking around in memory we load it a second time.  This
+                // will effectively just increment the refcount since it has been
+                // loaded before.
+                let handle_lock = LoadLibraryExW(
+                    module_path.as_ptr(),
+                    ptr::null_mut(),
+                    LOAD_LIBRARY_AS_DATAFILE,
+                );
+
+                let mut vmem_info = mem::zeroed();
+                let mut should_break = false;
+                if VirtualQuery(
+                    module_info.lpBaseOfDll,
+                    &mut vmem_info,
+                    mem::size_of::<MEMORY_BASIC_INFORMATION>(),
+                ) == mem::size_of::<MEMORY_BASIC_INFORMATION>()
+                {
+                    let module_path = OsString::from_wide(
+                        &module_path[..module_path.iter().position(|x| *x == 0).unwrap_or(0)],
```

Comment:
The return value of `GetModuleFileNameExW` should be this length. Not sure if we should be limiting it to that instead or as well.